### PR TITLE
[13.0] [MIG] quality_control_oca: Migration to v13.0

### PR DESCRIPTION
**Pull request by CarlosRoca13**
cc @Tecnativa TT26773

This PR superseds https://github.com/OCA/manufacture/pull/481 ping @jarroyomorales @ps-tubtim 

please @victoralmau @ernestotejeda review this

### CODE REVIEW

File: quality_control_oca/models/qc_inspection.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+        for inspection in self:
+            if not inspection.test:
+                raise exceptions.UserError(_("You must first set the test to perform."))
+        self.write({"state": "ready"})
+
+    def action_confirm(self):
+        for inspection in self:
+            for line in inspection.inspection_lines:
+                if line.question_type == "qualitative":
+                    if not line.qualitative_value:
```
Comment (ernestotejeda) [inline quality_control_oca/models/qc_inspection.py]: These 2 statements can be written in a single 'if' statement to save indentation levels

File: quality_control_oca/models/qc_test_category.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        ids = self.ids
+        level = 100
+        while ids:
+            parents = self.search([("id", "in", ids), ("parent_id", "!=", False)])
+            ids = list({x.parent_id.id for x in parents})
+            if not level:
+                raise exceptions.UserError(
+                    _("Error! You can not create recursive categories.")
+                )
+            level -= 1
```
Comment (ernestotejeda) [inline quality_control_oca/models/qc_test_category.py]: Maybe _check_recursion ORM method can be used here. what do you think?
```suggestion
    @api.constrains("parent_id")
    def _check_parent_id(self):
        if not self._check_recursion():
            raise ValidationError(_("Error! You can not create recursive categories."))
```

File: quality_control_oca/security/quality_control_security.xml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            eval="[(4, ref('base.user_root')), (4, ref('base.user_admin'))]"
+        />
+    </record>
+    <record id="qc_test_multi_company_rule" model="ir.rule">
+        <field name="name">Quality control test multi-company</field>
+        <field name="model_id" ref="model_qc_test" />
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >['|', ('company_id', '=', False), ('company_id', 'child_of', [user.company_id.id])]</field>
```
Comment (ernestotejeda) [inline quality_control_oca/security/quality_control_security.xml]:
```suggestion
        >['|', ('company_id', '=', False), ('company_id', 'in', company_ids)]</field>
```

File: quality_control_oca/security/quality_control_security.xml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    </record>
+    <record id="qc_inspection_multi_company_rule" model="ir.rule">
+        <field name="name">Quality control inspection multi-company</field>
+        <field name="model_id" ref="model_qc_inspection" />
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >['|', ('company_id', '=', False), ('company_id', 'child_of', [user.company_id.id])]</field>
```
Comment (ernestotejeda) [inline quality_control_oca/security/quality_control_security.xml]:
```suggestion
        >['|', ('company_id', '=', False), ('company_id', 'in', company_ids)]</field>
```

File: quality_control_oca/security/quality_control_security.xml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    </record>
+    <record id="qc_trigger_multi_company_rule" model="ir.rule">
+        <field name="name">Quality control trigger multi-company</field>
+        <field name="model_id" ref="model_qc_trigger" />
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >['|', ('company_id', '=', False), ('company_id', 'child_of', [user.company_id.id])]</field>
```
Comment (ernestotejeda) [inline quality_control_oca/security/quality_control_security.xml]:
```suggestion
        >['|', ('company_id', '=', False), ('company_id', 'in', company_ids)]</field>
```

File: quality_control_oca/views/qc_inspection_view.xml [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+                        type="object"
+                        attrs="{'invisible': [('state', 'not in', ['waiting', 'ready', 'failed', 'success'])]}"
+                        string="Cancel"
+                        icon="fa-ban"
+                    />
+                    <field
+                        name="state"
+                        widget="statusbar"
+                        statusbar_visible="draft,waiting,success"
+                        statusbar_colors='{"success": "blue", "failed": "red"}'
```
Comment (ernestotejeda) [inline quality_control_oca/views/qc_inspection_view.xml]: Use `decoration-{$name}` instead

Comment (CarlosRoca13) [reply quality_control_oca/views/qc_inspection_view.xml]: In v12 the colors were not applied to the status bar yet. So I choosed to delete this for keep the odoo style.

Comment (ernestotejeda) [reply quality_control_oca/views/qc_inspection_view.xml]: I'm sorry @CarlosRoca13, I was confused. But yes, I agree to remove it, I think it's an attribute used in old Odoo versions and now it doesn't work, it no longer exists (neither in v12 nor in v13)

**Review by ernestotejeda — APPROVED** [review]

File: quality_control_oca/data/quality_control_data.xml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        <record
+            forcecreate="True"
+            id="decimal_quality_control"
+            model="decimal.precision"
+        >
+            <field name="name">Quality Control</field>
+            <field name="digits">5</field>
+        </record>
+    </data>
+    <data>
```
Comment (victoralmau) [inline quality_control_oca/data/quality_control_data.xml]:
```suggestion
    <data noupdate="1">
```

File: quality_control_oca/models/qc_inspection.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+            ("ready", "Ready"),
+            ("waiting", "Waiting supervisor approval"),
+            ("success", "Quality success"),
+            ("failed", "Quality failed"),
+            ("canceled", "Canceled"),
+        ],
+        string="State",
+        readonly=True,
+        default="draft",
+        track_visibility="onchange",
```
Comment (victoralmau) [inline quality_control_oca/models/qc_inspection.py]:
```suggestion
        tracking=True,
```

File: quality_control_oca/models/qc_inspection.py [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+        comodel_name="res.company",
+        string="Company",
+        readonly=True,
+        states={"draft": [("readonly", False)]},
+        default=lambda self: self.env.company,
+    )
+    user = fields.Many2one(
+        comodel_name="res.users",
+        string="Responsible",
+        track_visibility="always",
```
Comment (victoralmau) [inline quality_control_oca/models/qc_inspection.py]:
```suggestion
        tracking=True,
```

File: quality_control_oca/models/qc_trigger_line.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+class QcTriggerLine(models.AbstractModel):
+    _name = "qc.trigger.line"
+    _description = "Abstract line for defining triggers"
+
+    trigger = fields.Many2one(comodel_name="qc.trigger", required=True)
+    test = fields.Many2one(comodel_name="qc.test", required=True)
+    user = fields.Many2one(
+        comodel_name="res.users",
+        string="Responsible",
+        track_visibility="always",
```
Comment (victoralmau) [inline quality_control_oca/models/qc_trigger_line.py]:
```suggestion
        tracking=True,
```

**Review by victoralmau — CHANGES_REQUESTED** [review]

**Comment by CarlosRoca13** [comment]
@victoralmau changes done

**Review by victoralmau — APPROVED** [review]

**Comment by pedrobaeza** [comment]
The module renaming was not done across all commits, so on next version, the commit history will be lost. I have amended that, and squash together the migration + renaming commit, using co-authoring feature for not having a big diff for no real gain.

Thanks to all the contributors.

/ocabot merge nobump

**Comment by SimoRubi** [comment]
Hi all, thanks for the PR!
May I ask why there is no need for a migration script here? I expected one because of the module renaming.
Maybe to avoid any clashing with core's `quality_control` module?

**Comment by pedrobaeza** [comment]
That's not done here, but on OpenUpgrade: https://github.com/OCA/OpenUpgrade/commit/63e67d0c235b002175c72b911aac8554fd37089c

**Pull request merged**